geode-core/src/main/java/org/apache/geode/management/internal/configuration/mutators/CacheConfigurationManager.java

```diff
@@ -0,0 +1,127 @@
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.geode.management.internal.configuration.mutators;
+
+import java.util.List;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import org.apache.logging.log4j.Logger;
+
+import org.apache.geode.annotations.Experimental;
+import org.apache.geode.cache.configuration.CacheConfig;
+import org.apache.geode.distributed.internal.InternalConfigurationPersistenceService;
+import org.apache.geode.logging.internal.log4j.api.LogService;
+import org.apache.geode.management.configuration.AbstractConfiguration;
+
+/**
+ * Defines the behavior to mutate a configuration change into a pre-existing cache config from a
+ * locator
+ * {@link org.apache.geode.distributed.ConfigurationPersistenceService}. Created with an object of
+ * type {@link CacheElement}, which represents the configuration change.
+ */
+@Experimental
+public abstract class CacheConfigurationManager<T extends AbstractConfiguration>
+    implements ConfigurationManager<T> {
+  private static final Logger logger = LogService.getLogger();
+
+  /**
+   * specify how to add the config to the existing cache config. Note at this point, the config
+   * should have passed all the validations already.
+   */
+  public abstract void add(T config, CacheConfig existing);
+
+  public abstract void update(T config, CacheConfig existing);
+
+  public abstract void delete(T config, CacheConfig existing);
+
+  public abstract List<T> list(T filterConfig, CacheConfig existing);
+
+  public abstract T get(String id, CacheConfig existing);
+
+  /**
+   * @param incoming the one that's about to be persisted
+   * @param group the group name of the existing cache element
+   * @param existing the existing cache element on another group
+   * @throws IllegalArgumentException if the incoming CacheElement is not compatible with the
+   *         existing
+   *
+   *         Note: incoming and existing should have the same ID already
+   */
+  public void checkCompatibility(T incoming, String group, T existing) {};
+
+  public final boolean add(InternalConfigurationPersistenceService service, T config,
+      String groupName) {
+    AtomicBoolean success = new AtomicBoolean(true);
+    service.updateCacheConfig(groupName, cacheConfigForGroup -> {
+      try {
+        add(config, cacheConfigForGroup);
```

Comment:
This is the only line that differs between `add()`, `delete()`, and `update()`. The rest is duplicated. Consider moving the duplicated code to a new method (maybe `applyOperation()` or something like that) that takes a group name and a `BiConsumer<? super T, UnaryOperator<CacheConfig>>`. Then all of this duplicated code boils down to:
```
applyOperation(groupName, this::add);
```